pyat/at/lattice/lattice_object.py

```diff
@@ -314,21 +306,175 @@ def _addition_filter(self, elems: Iterable[Element]):
+            if abs(periodicity - nbp) > _TWO_PI_ERROR:
+                warn(AtWarning('Non-integer number of cells: {}/{}. Periodi'
+                               'city set to 1'.format(self.periodicity, n)))
+                periodicity = 1
+        try:
+            cell_h = self._cell_harmnumber
+        except AttributeError:
+            hdict = {}
+        else:
+            hdict = dict(_cell_harmnumber=n*cell_h)
+        elems = (copy_fun(el, copy_elements) for _ in range(n) for el in self)
+        return Lattice(elem_generator, elems, iterator=self.attrs_filter,
+                       periodicity=periodicity, **hdict)
+
+    def concatenate(self, *lattices: Iterable[Element],
+                    copy_elements=False, copy=False):
+        """Concatenate several `Iterable[Element]` with the lattice
+
+        Equivalents syntaxes:
+        >>> newring = ring.concatenate(r1, r2, r3, copy=True)
+        >>> newring = ring + r1 + r2 + r3
+
+        >>> ring.concatenate(r1, copy=False)
+        >>> ring += r1
+
+        Parameters:
+            lattices: :py:obj:`Iterables[Element]` to be concatenanted
+                      to the Lattice, several lattices are allowed
+                      (see example)
+
+        Keyword Arguments:
+            copy_elements(bool): Default :py:obj:`False`. If :py:obj:`True`
+                        deepcopies of the elements of lattices are used
+            copy(bool): Default :py:obj:`False`. If :py:obj:`True`
+                           the lattice is modified in place.
+                           Oterwise a new Lattice object is returned
+
+        Returns:
+            lattice(Lattice): concatenated Lattice, if `copy==True` the
+                              new lattice object is returned
+                              otherwise None
+        """
+        if copy:
+            lattice = Lattice(self)
+        else:
+            lattice = self
+        for lat in lattices:
+            lattice.extend(lat, copy_elements=copy_elements)
+        return lattice if copy else None
+
+    def reverse(self, copy=False):
+        r"""Reverse the order of the lattice and swapt the faces
+        of elements. Alignment errors are not swapped
+
+
+        Equivalents syntaxes:
+        >>> newring = ring.reverse(copy=True)
+        >>> newring = reversed(ring)
+
+        Keyword Arguments:
+            copy(bool): Default :py:obj:`False`. If :py:obj:`True`
+                           the lattice is modified in place.
+                           Oterwise a new Lattice object is returned
+
+        Returns:
+            lattice(Lattice): reversed Lattice, if `copy==True` the
+                              new lattice object is returned
+                              otherwise None
+        """
+        if copy:
+            lattice = Lattice(self)
+        else:
+            lattice = self
+        lattice[:] = lattice[::-1]
+        for e in lattice:
+            e.swap_faces()
```

Comment:
There is a problem: even if `copy` is True, the faces of the original lattice are swapped !

You must make a copy of the element before swapping the faces.

Reply:
Or maybe the copy should be done if the swap_faces() method ? And only if necessary (if there is anything to swap?)